cpp/include/cuvs/neighbors/ann_mg.hpp

```diff
@@ -0,0 +1,1417 @@
+                                      int64_t n_cols,
+                                      int64_t n_neighbors,
+                                      int64_t n_batches) const;
+
+  distribution_mode mode_;
+  int num_ranks_;
+  std::vector<ann_interface<AnnIndexType, T, IdxT>> ann_interfaces_;
+};
+
+/// \defgroup ann_mg_cpp_index_build ANN MG index build
+
+/// \ingroup ann_mg_cpp_index_build
+/**
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   raft::handle_t handle;
+ *   cuvs::neighbors::mg::nccl_clique& clique; // default NCCL clique
+ *   cuvs::neighbors::ivf_flat::mg_index_params index_params; // default build parameters
+ *   auto index = cuvs::neighbors::mg::build(handle, clique, index_params, index_dataset);
+ * @endcode
+ *
+ * @param[in] handle
+ * @param[in] clique configure the NCCL clique
+ * @param[in] index_params configure the index building
+ * @param[in] index_dataset a row-major matrix on host [n_rows, dim]
+ *
+ * @return the constructed IVF-Flat MG index
+ */
+auto build(const raft::resources& handle,
+           const cuvs::neighbors::mg::nccl_clique& clique,
+           const ivf_flat::mg_index_params& index_params,
+           raft::host_matrix_view<const float, int64_t, row_major> index_dataset)
+  -> ann_mg_index<ivf_flat::index<float, int64_t>, float, int64_t>;
+
+/// \ingroup ann_mg_cpp_index_build
+/**
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   raft::handle_t handle;
+ *   cuvs::neighbors::mg::nccl_clique& clique; // default NCCL clique
+ *   cuvs::neighbors::ivf_flat::mg_index_params index_params; // default build parameters
+ *   auto index = cuvs::neighbors::mg::build(handle, clique, index_params, index_dataset);
+ * @endcode
+ *
+ * @param[in] handle
+ * @param[in] clique configure the NCCL clique
+ * @param[in] index_params configure the index building
+ * @param[in] index_dataset a row-major matrix on host [n_rows, dim]
+ *
+ * @return the constructed IVF-Flat MG index
+ */
+auto build(const raft::resources& handle,
+           const cuvs::neighbors::mg::nccl_clique& clique,
+           const ivf_flat::mg_index_params& index_params,
+           raft::host_matrix_view<const int8_t, int64_t, row_major> index_dataset)
+  -> ann_mg_index<ivf_flat::index<int8_t, int64_t>, int8_t, int64_t>;
+
+/// \ingroup ann_mg_cpp_index_build
+/**
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   raft::handle_t handle;
+ *   cuvs::neighbors::mg::nccl_clique& clique; // default NCCL clique
+ *   cuvs::neighbors::ivf_flat::mg_index_params index_params; // default build parameters
+ *   auto index = cuvs::neighbors::mg::build(handle, clique, index_params, index_dataset);
+ * @endcode
+ *
+ * @param[in] handle
+ * @param[in] clique configure the NCCL clique
+ * @param[in] index_params configure the index building
+ * @param[in] index_dataset a row-major matrix on host [n_rows, dim]
+ *
+ * @return the constructed IVF-Flat MG index
+ */
+auto build(const raft::resources& handle,
+           const cuvs::neighbors::mg::nccl_clique& clique,
```

Comment:
If we can remove this argument then we can match the single gpu APIs. I think I'd prefer to put this argument on the resource handle. Please create a new resource type for the nccl clique like we do in cuml.